File: internal/content/testdata/bench/java/allocate-books.java
import java.util.*;

class Solution {
    public long allocate_books(long[] pages, long students) {
        int n = pages.length;
        if (students > n) return -1;
        long lo = 0, hi = 0;
        for (long p : pages) {
            lo = Math.max(lo, p);
            hi += p;
        }
        long ans = hi;
        while (lo <= hi) {
            long mid = (lo + hi) / 2;
            if (canAllocate(pages, students, mid)) {
                ans = mid;
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }
        return ans;
    }

    private boolean canAllocate(long[] pages, long students, long maxPages) {
        long count = 1, total = 0;
        for (long p : pages) {
            if (p > maxPages) return false;
            if (total + p > maxPages) {
                count++;
                total = p;
            } else {
                total += p;
            }
        }
        return count <= students;
    }
}

File: internal/content/testdata/bench/rust/allocate-books.rs
fn allocate_books(pages: Vec<i64>, students: i64) -> i64 {
    let n = pages.len() as i64;
    if students > n {
        return -1;
    }
    fn can_allocate(pages: &Vec<i64>, students: i64, max_pages: i64) -> bool {
        let mut count: i64 = 1;
        let mut total: i64 = 0;
        for &p in pages {
            if p > max_pages {
                return false;
            }
            if total + p > max_pages {
                count += 1;
                total = p;
            } else {
                total += p;
            }
        }
        count <= students
    }
    let mut lo = *pages.iter().max().unwrap();
    let mut hi: i64 = pages.iter().sum();
    let mut ans = hi;
    while lo <= hi {
        let mid = (lo + hi) / 2;
        if can_allocate(&pages, students, mid) {
            ans = mid;
            hi = mid - 1;
        } else {
            lo = mid + 1;
        }
    }
    ans
}
